File: tutorials/01-s3/node/s3-demo.mjs
/**
 * S3 against Floci, using AWS SDK v3.
 *
 * Behaviourally identical to ../python/s3_demo.py -- read whichever language
 * you know and map it onto the other.
 *
 *   npm install
 *   node s3-demo.mjs
 */

import {
  S3Client,
  CreateBucketCommand,
  PutObjectCommand,
  GetObjectCommand,
  ListObjectsV2Command,
  PutBucketVersioningCommand,
  ListObjectVersionsCommand,
  DeleteObjectCommand,
  DeleteBucketCommand,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

// ---------------------------------------------------------------------------
// The ONLY Floci-specific lines in this file.
//
// Drop `endpoint` and `forcePathStyle` and this runs against real AWS with
// whatever credentials the environment provides.
//
// forcePathStyle matters locally: real S3 uses virtual-host addressing
// (bucket.s3.amazonaws.com), which cannot work against localhost.
// ---------------------------------------------------------------------------
const s3 = new S3Client({
  endpoint: process.env.AWS_ENDPOINT_URL ?? "http://localhost:4566",
  region: process.env.AWS_DEFAULT_REGION ?? "us-east-1",
  forcePathStyle: true,
  credentials: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID ?? "test",
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY ?? "test",
  },
});

const BUCKET = "floci-demo-node";
const KEY = "notes/sample.txt";

const step = (msg) => console.log(`\n==> ${msg}`);

async function main() {
  step(`Creating bucket ${BUCKET}`);
  try {
    await s3.send(new CreateBucketCommand({ Bucket: BUCKET }));
  } catch (err) {
    if (err.name !== "BucketAlreadyOwnedByYou") throw err;
    console.log("    already exists, reusing it");
  }

  step("Uploading an object");
  await s3.send(
    new PutObjectCommand({ Bucket: BUCKET, Key: KEY, Body: "the quick brown fox\n" })
  );

  step("Listing with a prefix");
  const listing = await s3.send(
    new ListObjectsV2Command({ Bucket: BUCKET, Prefix: "notes/" })
  );
  for (const obj of listing.Contents ?? []) {
    console.log(`    ${obj.Key}  (${obj.Size} bytes)`);
  }

  step("Reading it back");
  const got = await s3.send(new GetObjectCommand({ Bucket: BUCKET, Key: KEY }));
  console.log(`    ${JSON.stringify((await got.Body.transformToString()).trim())}`);

  step("Generating a presigned URL (valid 5 minutes)");
  const url = await getSignedUrl(
    s3,
    new GetObjectCommand({ Bucket: BUCKET, Key: KEY }),
    { expiresIn: 300 }
  );
  console.log(`    ${url}`);

  step("Fetching that URL with no credentials at all");
  // Plain fetch, no SDK, no credentials. The query-string signature is the
  // entire authorisation.
  const res = await fetch(url);
  console.log(`    HTTP ${res.status}: ${JSON.stringify((await res.text()).trim())}`);

  step("Enabling versioning and overwriting");
  await s3.send(
    new PutBucketVersioningCommand({
      Bucket: BUCKET,
      VersioningConfiguration: { Status: "Enabled" },
    })
  );
  await s3.send(
    new PutObjectCommand({ Bucket: BUCKET, Key: KEY, Body: "jumps over the lazy dog\n" })
  );

  const { Versions = [], DeleteMarkers = [] } = await s3.send(
    new ListObjectVersionsCommand({ Bucket: BUCKET, Prefix: "notes/" })
  );
  for (const v of Versions) {
    console.log(`    ${v.Key}  version=${v.VersionId}${v.IsLatest ? "  <- current" : ""}`);
  }

  step("Cleaning up");
  // A versioned bucket needs every version removed before it will delete.
  for (const v of [...Versions, ...DeleteMarkers]) {
    await s3.send(
      new DeleteObjectCommand({ Bucket: BUCKET, Key: v.Key, VersionId: v.VersionId })
    );
  }
  await s3.send(new DeleteBucketCommand({ Bucket: BUCKET }));
  console.log("    done");
}

main().catch((err) => {
  console.error("\nFailed:", err.message);
  process.exit(1);
});

File: tutorials/01-s3/python/s3_demo.py
"""
S3 against Floci, using boto3.

Does the same sequence as the CLI walkthrough in ../README.md:
create -> upload -> presign -> fetch anonymously -> version -> clean up.

    pip install -r requirements.txt
    python s3_demo.py
"""

import os
import urllib.request

import boto3

# ---------------------------------------------------------------------------
# The ONLY Floci-specific line in this file.
#
# Delete endpoint_url and this exact script runs against real AWS, using
# whatever credentials your environment already has. Everything below is
# ordinary boto3.
# ---------------------------------------------------------------------------
ENDPOINT = os.environ.get("AWS_ENDPOINT_URL", "http://localhost:4566")

s3 = boto3.client(
    "s3",
    endpoint_url=ENDPOINT,
    region_name=os.environ.get("AWS_DEFAULT_REGION", "us-east-1"),
    aws_access_key_id=os.environ.get("AWS_ACCESS_KEY_ID", "test"),
    aws_secret_access_key=os.environ.get("AWS_SECRET_ACCESS_KEY", "test"),
)

BUCKET = "floci-demo-python"
KEY = "notes/sample.txt"


def step(msg):
    print(f"\n==> {msg}")


def main():
    step(f"Creating bucket {BUCKET}")
    try:
        s3.create_bucket(Bucket=BUCKET)
    except s3.exceptions.BucketAlreadyOwnedByYou:
        print("    already exists, reusing it")

    step("Uploading an object")
    s3.put_object(Bucket=BUCKET, Key=KEY, Body=b"the quick brown fox\n")

    step("Listing with a prefix")
    listing = s3.list_objects_v2(Bucket=BUCKET, Prefix="notes/")
    for obj in listing.get("Contents", []):
        print(f"    {obj['Key']}  ({obj['Size']} bytes)")

    step("Reading it back")
    body = s3.get_object(Bucket=BUCKET, Key=KEY)["Body"].read().decode()
    print(f"    {body.strip()!r}")

    step("Generating a presigned URL (valid 5 minutes)")
    url = s3.generate_presigned_url(
        "get_object",
        Params={"Bucket": BUCKET, "Key": KEY},
        ExpiresIn=300,
    )
    print(f"    {url}")

    step("Fetching that URL with no credentials at all")
    # urllib knows nothing about AWS. The signature in the query string is the
    # entire authorisation -- that is the point of a presigned URL.
    with urllib.request.urlopen(url, timeout=10) as resp:
        print(f"    HTTP {resp.status}: {resp.read().decode().strip()!r}")

    step("Enabling versioning and overwriting")
    s3.put_bucket_versioning(
        Bucket=BUCKET,
        VersioningConfiguration={"Status": "Enabled"},
    )
    s3.put_object(Bucket=BUCKET, Key=KEY, Body=b"jumps over the lazy dog\n")

    versions = s3.list_object_versions(Bucket=BUCKET, Prefix="notes/").get("Versions", [])
    for v in versions:
        marker = "  <- current" if v["IsLatest"] else ""
        print(f"    {v['Key']}  version={v['VersionId']}{marker}")

    step("Cleaning up")
    # A versioned bucket needs every version removed before it will delete.
    for v in versions:
        s3.delete_object(Bucket=BUCKET, Key=v["Key"], VersionId=v["VersionId"])
    for m in s3.list_object_versions(Bucket=BUCKET).get("DeleteMarkers", []):
        s3.delete_object(Bucket=BUCKET, Key=m["Key"], VersionId=m["VersionId"])
    s3.delete_bucket(Bucket=BUCKET)
    print("    done")


if __name__ == "__main__":
    main()
